Userland/Libraries/LibWeb/CSS/StyleProperties.cpp

```diff
@@ -159,38 +152,89 @@ void StyleProperties::load_font() const
             size = 12;
             break;
         case CSS::ValueID::Smaller:
-            // FIXME: This should be relative to the parent.
-            size = 10;
+            size = roundf(parent_font_size / font_size_ratio);
             break;
         case CSS::ValueID::Larger:
-            // FIXME: This should be relative to the parent.
-            size = 12;
+            size = roundf(parent_font_size * font_size_ratio);
             break;
 
         default:
             break;
         }
-    } else if (font_size->is_length()) {
-        // FIXME: This isn't really a length, it's a numeric value..
-        int font_size_integer = font_size->to_length().raw_value();
-        size = font_size_integer;
+    } else {
+        Optional<Length> maybe_length;
+        if (font_size->is_length()) {
+            maybe_length = font_size->to_length();
+        } else if (font_size->is_calculated()) {
+            Length length = Length(0, Length::Type::Calculated);
+            length.set_calculated_style(verify_cast<CalculatedStyleValue>(font_size.ptr()));
+            maybe_length = length;
+        }
+        if (maybe_length.has_value()) {
+            // FIXME: em sizes return 0 here, for some reason
+            auto calculated_size = maybe_length.value().resolved_or_zero(node, parent_font_size).to_px(node);
+            if (calculated_size != 0)
+                size = calculated_size;
+        }
     }
 
-    FontSelector font_selector { family, size, weight };
+    // Note: Both of these are modified by the find_font() lambda
+    FontSelector font_selector;
+    bool monospace = false;
 
-    auto found_font = FontCache::the().get(font_selector);
-    if (found_font) {
-        m_font = found_font;
-        return;
-    }
+    auto find_font = [&](String const& family, bool is_monospace = false) -> RefPtr<Gfx::Font> {
+        font_selector = { family, size, weight };
+        monospace = is_monospace;
 
-    Gfx::FontDatabase::the().for_each_font([&](auto& font) {
-        if (font.family() == family && font.weight() == weight && font.presentation_size() == size)
-            found_font = font;
-    });
+        if (auto found_font = FontCache::the().get(font_selector))
+            return found_font;
+
+        if (auto found_font = Gfx::FontDatabase::the().get(family, size, weight))
+            return found_font;
+
+        return {};
+    };
+
+    auto find_generic_font = [&](ValueID font_id) -> RefPtr<Gfx::Font> {
+        switch (font_id) {
```

Comment:
I don't like the hardcoded values here...
To get the default system monospace and variable-width fonts you can query them via `FontDatabase::default_font()` and `FontDatabase::default_fixed_width_font()`. If there are more default font to support in the future, they should also get a static getter from the FontDatabase, and not get specified here imo.

Reply:
Very good point. I should have looked around in FontDatabase some more too, I think the code here was written before most of that was implemented.

Reply:
Turns out this doesn't work - `FontDatabase::default_font()` and friends all return a single *font*, which is a typeface at a specific size and weight. What we want is just the typeface name. I could split the string returned by `FontDatabase::default_font_query()`, but that sounds very error prone.